staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_patch.go

```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+func (lw *cacherListerWatcher) kcpAwareResourcePrefix() string {
+	if lw.kcpExtraStorageMetadata.Cluster.Wildcard {
+		return lw.resourcePrefix
+	}
+
+	// This is a request for normal (non-bound) CRs outside of system:system-crds. Make sure we only list in the
+	// specific logical cluster.
+	return lw.resourcePrefix + "/" + lw.kcpExtraStorageMetadata.Cluster.Name.String()
```

Comment:
I don't understand why we need change the `resourcePrefix`. 

My understanding is that the cluster name will by set by `createKCPClusterAwareContext` function and then used by https://github.com/kcp-dev/kubernetes/blob/e505b8d60dc9ede062b0c38dedbef03a696e9f83/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L390 to calculate the key for etcd

Reply:
If we don't do this, we have the conversion issue

Reply:
OK, I think this is because the reflector used by the watch-cache gets `config.Storage` (etcd) interface which expects the key. It doesn't use the generic storage interface (linked above).

Reply:
Yes we're going direct to etcd here.

Reply:
This is a get call with a prefix of .../customresources unless we change it (for normal crds). If we don't change it, a watch cache instance for v1 widgets gets all versions when listing and watching, leading to the conversion issue

Reply:
would it make sense to use https://github.com/kcp-dev/kubernetes/blob/e505b8d60dc9ede062b0c38dedbef03a696e9f83/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L84 ? any benefit ? I think this is the KeyFunc from the generic registry.

Reply:
We don't have individual runtime objects. We are just doing a list and have to specify the prefix

Reply:
ok, thx